lib/property_test.ex

```diff
@@ -191,9 +167,48 @@ defmodule PropertyTest do
       case StreamData.check_all(property, options, &(&1.())) do
         {:ok, _result} ->
           :ok
-        {:error, result} ->
-          raise Error, result
+        {:error, test_result} ->
+          PropertyTest.__raise__(test_result)
       end
     end
   end
+
+  def __raise__(test_result) do
+    %{original_failure: original_failure,
+      shrunk_failure: shrunk_failure,
+      nodes_visited: nodes_visited} = test_result
+    choose_error_and_raise(original_failure, shrunk_failure, nodes_visited)
+  end
+
+  defp choose_error_and_raise(_, %{exception: %AssertionError{}} = shrunk_failure, nodes_visited) do
+    reraise enrich_assertion_error(shrunk_failure, nodes_visited), shrunk_failure.stacktrace
+  end
+
+  defp choose_error_and_raise(%{exception: %AssertionError{}} = original_failure, _, nodes_visited) do
+    reraise enrich_assertion_error(original_failure, nodes_visited), original_failure.stacktrace
+  end
+
+  defp choose_error_and_raise(_original_failure, shrunk_failure, _nodes_visited) do
+    formatted = indent(Exception.format(:error, shrunk_failure.exception, shrunk_failure.stacktrace))
+    message = "failed with generated values:\n\n#{format_generated_values(shrunk_failure.generated_values)}\n\n#{formatted}"
+    raise Error, message: message
+  end
+
+  defp enrich_assertion_error(%{exception: exception, generated_values: generated_values}, _nodes_visited) do
+    message =
+      "Failed with generated values:\n\n#{format_generated_values(generated_values)}" <>
+      if(is_binary(exception.message), do: "\n\n" <> exception.message, else: "")
+
+    %{exception | message: message}
+  end
+
+  defp format_generated_values(values) do
```

Comment:
My only thought about improving it would be to track variables but I don't think it is important to be quite honest. I just wanted to understand the process.

Btw, can you indent the code with four spaces? To keep aligned with markdown?